File: examples/AskDemo.scala
import layoutz._

object AskDemo {
  def main(args: Array[String]): Unit = {
    print("[2J[H")

    def gap(): Unit = println()

    val companions = Seq(
      "Bilbo",
      "Balin",
      "Dwalin",
      "Thorin",
      "Gandalf",
      "Kili",
      "Fili",
      "Bombur",
      "Bofur",
      "Gloin",
      "Oin",
      "Dori"
    )

    val realms = Seq("The Shire", "Rivendell", "Mirkwood", "Lake-town", "Erebor")

    Ask.input("Name › ", placeholder = "anonymous")
    gap()

    Ask.choose("Choose a realm", realms)
    gap()

    Ask.chooseMany(
      "Pack provisions (up to 3)",
      Seq("lembas", "pipe-weed", "waybread", "miruvor", "rope", "athelas"),
      limit = 3
    )
    gap()

    Ask.filter("Search a companion › ", companions)
    gap()

    Ask.write(prompt = "Pose a riddle", placeholder = "This thing all things devours…")
    gap()

    Ask.file(start = "demos")
    gap()

    Ask.confirm("Venture on the quest?", default = true)
    gap()

    Ask.pager(
      (0 until 40)
        .map { i =>
          val c = companions(i % companions.length)
          f"${i + 1}%2d.  $c%-12s ${realms(i % realms.length)}"
        }
        .mkString("\n"),
      height = 8
    )
    gap()

    Ask.spin("Awaking Smaug…") {
      Thread.sleep(1500)
      "ready"
    }
  }
}

File: layoutz-hs/examples/AskDemo.hs
{-# LANGUAGE OverloadedStrings #-}

{-|
Ask demo - a tour of the one-shot @ask*@ prompts. Each takes over the tty
briefly, returns a value, and leaves a single committed line behind.

Run with: cabal run ask-demo
-}

module Main where

import Layoutz
import Control.Concurrent (threadDelay)
import Control.Monad (void)
import Text.Printf (printf)

companions :: [String]
companions =
  [ "Bilbo", "Balin", "Dwalin", "Thorin"
  , "Gandalf", "Kili", "Fili", "Bombur"
  , "Bofur", "Gloin", "Oin", "Dori"
  ]

realms :: [String]
realms = ["The Shire", "Rivendell", "Mirkwood", "Lake-town", "Erebor"]

gap :: IO ()
gap = putStrLn ""

main :: IO ()
main = do
  putStr "\ESC[2J\ESC[H"

  _ <- askInput "Name › " "anonymous" ""
  gap
  _ <- askChoose "Choose a realm" realms id
  gap
  _ <- askChooseMany "Pack provisions (up to 3)"
         ["lembas", "pipe-weed", "waybread", "miruvor", "rope", "athelas"] 3 id
  gap
  _ <- askFilter "Search a companion › " companions 8 id
  gap
  _ <- askWrite "Pose a riddle" "This thing all things devours…" "" "Ctrl-D to save"
  gap
  _ <- askFile "." 12
  gap
  _ <- askConfirm "Venture on the quest?" True "Yes" "No"
  gap
  askPager roster 8 True
  gap
  void $ askSpin "Awaking Smaug…" SpinnerDots (threadDelay 1500000 >> pure ("ready" :: String))
  where
    roster = unlines
      [ printf "%2d.  %-12s %s"
          (i + 1) (companions !! (i `mod` length companions))
          (realms !! (i `mod` length realms))
      | i <- [0 .. 39 :: Int] ]
